### PR TITLE
SERENE-869: Enable dynamic URI versioning

### PR DESCRIPTION
**Pull request by gaj67**

### CODE REVIEW

**Comment by gaj67** [comment]
In previous tests, the ordering of results via getAllEpgs() has matched the order of creation. This has now started to fail, possibly because status changes to COMPLETE (or ERROR) can occur asynchronously. Possibly we might give thought to ordering the SQL query by creation-time.

File: search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        if (!lookupPath.startsWith(API_VERSION_PREFIX)) {
+            logger.debug("Not versioned with our prefix!");
+            return null; // Not a versioned URI.
+        }
+        String[] subPaths = lookupPath.substring(API_VERSION_PREFIX.length()).split("/", 2);
+        final String restOfPath = (subPaths.length == 2) ? ("/" + subPaths[1]) : "";
+        // XXX Allow for backward compatibility - if using old 1.0 version, then
+        // remap to 1 and then increment to cause a retry.
+        int version = "1.0".equals(subPaths[0]) ? 2 : getVersion(subPaths[0]);
```
Comment (bowernj) [inline search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: This works for "v1", "v1.0", "v2", but it also works for any "/api/v*" request. Is that what we want?

ie. GET "/api/v5/indexes/" will eventually dec down to the v1 endpoint. Should there be an additional check for a current maximum version or something?

File: search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+                @Override
+                public String getServletPath() {
+                    return newPath;
+                }
+            };
+            logger.debug("Trying super lookup for path " + newPath + "...");
+            method = super.lookupHandlerMethod(newPath, newRequest);
+        }
+        logger.debug("Found method!");
```
Comment (bowernj) [inline search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: Minor: This debug statement will still log if method is null, eg. invalid API lookups

File: search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java (hunk 1 above)
Comment (bowernj) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: Do we actually want to support "v2" on all endpoints starting from now?

Comment (gaj67) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: Point 8 in the ticket description: it supports fallback to previous versions of endpoints if no change was made.

File: search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java (hunk 2 above)
Comment (gaj67) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: Good catch!

File: search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java (hunk 1 above)
Comment (gaj67) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: I don't think we can reliably know what a maximum version number might be.

Comment (gaj67) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: How did the Swagger API look?

Comment (bowernj) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: Swagger looks fine, shows "v1" only

Comment (gaj67) [reply search/src/main/java/au/csiro/data61/stellar/search/versioning/VersionRequestMappingHandlerMapping.java]: I fixed the debug problem you noted. Is this branch now ready to go?

**Pull request merged**